network/gossip/libp2p/topology/topicBasedTopology.go

```diff
@@ -0,0 +1,84 @@
+package topology
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// TopicBasedTopology is a deterministic topology mapping that creates a connected graph component among the nodes
+// involved in each topic.
+type TopicBasedTopology struct {
+	seed  int64                  // used for sampling connected graph
+	me    flow.Identifier        // used to keep identifier of the node
+	state protocol.ReadOnlyState // used to keep a read only protocol state
+}
+
+// NewTopicBasedTopology returns an instance of the TopicBasedTopology.
+func NewTopicBasedTopology(nodeID flow.Identifier, state protocol.ReadOnlyState) (*TopicBasedTopology, error) {
+	seed, err := seedFromID(nodeID)
```

Comment:
There is no guarantee currently that node IDs are random. Depending on what kind of properties we need from this seed it might be better to use something else. 

And just checking: it's OK that all the nodes use different seeds?

Reply:
Currently, we are following a deterministic approach, we need "some entropy" for load distribution, but not a very high entropy required. We do not formalize the connectedness as a function of entropy yet, so it is currently fine. 

>And just checking: it's OK that all the nodes use different seeds?

For the purpose of topology construction (even in randomized mode) as long as all nodes follow the same level of entropy it would be fine (would be happy to discuss it further).